### PR TITLE
✨ Make apigen a standalone go module

### LINKED ISSUE #2668: bug: can not `go install` apigen
### Describe the bug

As an edge-mc developer, I tried to `go install` kcp's `apigen` command, and got an error message that I do not understand.  Following is what it looked like.

```
(base) mspreitz@mjs12 ~ % mkdir test4

(base) mspreitz@mjs12 ~ % cd test4

(base) mspreitz@mjs12 test4 % GOBIN=$PWD go install github.com/kcp-dev/kcp/cmd/apigen@v0.10.1-0.20230119143945-4506fdc06406
go: github.com/kcp-dev/kcp/cmd/apigen@v0.10.1-0.20230119143945-4506fdc06406 (in github.com/kcp-dev/kcp@v0.10.1-0.20230119143945-4506fdc06406):
	The go.mod file for the module providing named packages contains one or
	more replace directives. It must not contain directives that would cause
	it to be interpreted differently than if it were the main module.

(base) mspreitz@mjs12 test4 % echo $?
1

(base) mspreitz@mjs12 test4 % go version
go version go1.19.5 darwin/amd64
```

### Steps To Reproduce

Shown above.

### Expected Behaviour

Successful `go install`.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ncdc**
## Summary
Make apigen a standalone go module so it can be installed by itself.

## Related issue(s)

Fixes #2668

### CODE REVIEW

File: hack/verify-go-modules.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+REPO_ROOT=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
+
+mapfile -t DIRS < <(find "${REPO_ROOT}" -name go.mod -print0 | xargs -0 dirname)
+
+for dir in "${DIRS[@]}"; do
+  pushd "$dir" &> /dev/null
```
Comment (MikeSpreitzer) [inline hack/verify-go-modules.sh:28]: Using `( ... )` is safer than `pushd $dir; ...; popd`

**Review by MikeSpreitzer — COMMENTED** [review]
I do not know how to test that it solves the problem.  I tried and failed as follows.

```
(base) mspreitz@mjs12 test5 % go install github.com/ncdc/kcp/cmd/apigen@bd5342130765ecec91053b5690e9e246c41372df
go: downloading github.com/ncdc/kcp/cmd/apigen v0.0.0-20230124185555-bd5342130765
go: downloading github.com/ncdc/kcp v0.0.0-20230124185555-bd5342130765
go: github.com/ncdc/kcp/cmd/apigen@bd5342130765ecec91053b5690e9e246c41372df: github.com/ncdc/kcp/cmd/apigen@v0.0.0-20230124185555-bd5342130765: parsing go.mod:
	module declares its path as: github.com/kcp-dev/kcp/cmd/apigen
	        but was required as: github.com/ncdc/kcp/cmd/apigen
```

**Comment by MikeSpreitzer** [comment]
I see that this avoids the "replace" for apis by requiring a specific version rather than replacing with the version in the same commit.  This requires updating the "require" line every time the apis change.  Just checking that this consequence is understood and approved by others.

File: hack/update-go-modules.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+# This script tidies all go modules in the repository.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+REPO_ROOT=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)
+
+mapfile -t DIRS < <(find "${REPO_ROOT}" -name go.mod -print0 | xargs -0 dirname)
```
Comment (MikeSpreitzer) [inline hack/update-go-modules.sh:25]: `dirname` terminates its output with a newline, right?  So despite all the effort, this still fails when a directory name includes a newline, right?  There are simpler approaches that accomplish as much.  OTOH, if you want to succeed when the directory name contains a newline, perhaps `find ... -exec process.sh \{\} \;`.

**Comment by ncdc** [comment]
> I do not know how to test that it solves the problem. I tried and failed as follows.

This will only work after it merges.

> I see that this avoids the "replace" for apis by requiring a specific version rather than replacing with the version in the same commit. This requires updating the "require" line every time the apis change. Just checking that this consequence is understood and approved by others.

We only have to update if/when the code in apigen's main.go needs to react to changes to the apis such that the generation logic needs to change. For typical api modifications, we shouldn't need to change anything.

File: hack/update-go-modules.sh (hunk 2 above)
Comment (ncdc) [reply hack/update-go-modules.sh:25]: I am following the shellcheck suggestions

**Comment by MikeSpreitzer** [comment]
Thanks!

**Comment by MikeSpreitzer** [comment]
Still having trouble testing.

```
(base) mspreitz@mjs12 ~ % mkdir test6

(base) mspreitz@mjs12 ~ % cd test6

(base) mspreitz@mjs12 test6 % GOBIN=$PWD go install github.com/ncdc/kcp/cmd/apigen@cb8dd3c46470f30414c6789be956162ccb21f669
go: github.com/ncdc/kcp/cmd/apigen@cb8dd3c46470f30414c6789be956162ccb21f669: github.com/ncdc/kcp/cmd/apigen@cb8dd3c46470f30414c6789be956162ccb21f669: invalid version: unknown revision cb8dd3c46470f30414c6789be956162ccb21f669


(base) mspreitz@mjs12 test6 % GOBIN=$PWD go install github.com/ncdc/kcp/cmd/apigen@v0.10.1-0.20230124223207-cb8dd3c46470  
go: github.com/ncdc/kcp/cmd/apigen@v0.10.1-0.20230124223207-cb8dd3c46470: github.com/ncdc/kcp/cmd/apigen@v0.10.1-0.20230124223207-cb8dd3c46470: invalid version: unknown revision cb8dd3c46470
```

**Comment by ncdc** [comment]
It looks like you're still trying to use my `ncdc` fork. You'll want to try github.com/kcp-dev/kcp/cmd/apigen instead:

```shell
$ cd /tmp
$ mkdir foo
$ cd foo
$ go install github.com/kcp-dev/kcp/cmd/apigen@main
go: downloading github.com/kcp-dev/kcp/cmd/apigen v0.0.0-20230124223207-cb8dd3c46470
go: downloading github.com/kcp-dev/kcp v0.10.1-0.20230124223207-cb8dd3c46470
```

**Comment by MikeSpreitzer** [comment]
D'oh!  Yes, it works when I do it right.

```
(base) mspreitz@mjs12 ~ % mkdir test7

(base) mspreitz@mjs12 ~ % cd test7

(base) mspreitz@mjs12 test7 % GOBIN=$PWD go install github.com/kcp-dev/kcp/cmd/apigen@cb8dd3c46470f30414c6789be956162ccb21f669
go: downloading github.com/kcp-dev/kcp/cmd/apigen v0.0.0-20230124223207-cb8dd3c46470
go: downloading github.com/kcp-dev/kcp/pkg/apis v0.10.1-0.20230123155139-7b3126cbf96c
go: downloading github.com/bombsimon/logrusr/v3 v3.1.0
go: downloading k8s.io/apiserver v0.24.3
go: downloading k8s.io/component-base v0.24.3

(base) mspreitz@mjs12 test7 % ls -l
total 111256
-rwxr-xr-x  1 mspreitz  staff  56960512 Jan 24 23:53 apigen

(base) mspreitz@mjs12 test7 % date
Tue Jan 24 23:54:08 EST 2023
```

File: hack/update-go-modules.sh (hunk 2 above)
Comment (MikeSpreitzer) [reply hack/update-go-modules.sh:25]: The following make shellcheck happy and works with newlines in the directory name.  It starts with a listing showing all the paths used in this test.

```bash-5.1$ find . -name go.mod
./ta
ste/go.mod
./go.mod
./docs/go.mod
./te  st/go.mod
./pkg/apis/go.mod

bash-5.1$ cat > process.sh
#!/usr/bin/env bash
echo "Working on (${1%/go.mod})"

bash-5.1$ chmod a+x process.sh

bash-5.1$ find . -name go.mod -exec "$PWD/process.sh" \{\} \;
Working on (./ta
ste)
Working on (.)
Working on (./docs)
Working on (./te  st)
Working on (./pkg/apis)
```

**Pull request merged**